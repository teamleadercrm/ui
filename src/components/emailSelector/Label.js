import React, { useCallback, useState, useEffect, useRef, useMemo } from 'react';
import PropTypes from 'prop-types';
import uiTypography from '@teamleader/ui-typography';
import cx from 'classnames';

import theme from './theme.css';
import { filterSuggestions, moveSuggestion, selectContentEditable, validateSuggestion } from './utils';
import Tag from '../tag';
import Link from '../link';
import Box from '../box';
import Autocomplete from './Autocomplete';
import Overlay from '../overlay';

const ENTER = 'Enter';
const ESCAPE = 'Escape';
const BACKSPACE = 'Backspace';
const UP_ARROW = 'ArrowUp';
const DOWN_ARROW = 'ArrowDown';
const SEMI = ';';
const COMMA = ',';

const Label = ({
  option,
  suggestions,
  editing,
  invalid,
  index,
  onClick,
  onFocus,
  onBlur,
  onRemove,
  onFinish,
  renderSuggestion,
}) => {
  const [content, setContent] = useState(option.email);
  const [autocompleteOpen, setAutocompleteOpen] = useState(true);
  const ref = useRef();

  const validSuggestions = useMemo(() => filterSuggestions(content, suggestions), [content, suggestions]);
  const [selectedSuggestion, setSelectedSuggestion] = useState(null);

  const hasValidSuggestions =
    autocompleteOpen && (Array.isArray(validSuggestions) ? validSuggestions : Object.keys(validSuggestions)).length > 0;

  useEffect(() => {
    if (editing) {
      setContent(option.email);
      selectContentEditable(ref.current);
    }
  }, [editing, option]);

  const onTagClick = useCallback(
    (event) => {
      onClick && onClick(index);
      event.stopPropagation();
      event.preventDefault();
    },
    [index, onClick],
  );

  const onTagFocus = useCallback(() => {
    onFocus && onFocus(index);
  }, [index, onFocus]);

  const onTagRemove = useCallback(
    (event) => {
      onRemove && onRemove(index);
      event.stopPropagation();
      event.preventDefault();
    },
    [index, onRemove],
  );

  const onChange = useCallback(
    (event) => {
      setContent(event.target.innerText);
    },
    [setContent],
  );

  const onTagBlur = useCallback(
    (event) => {
      const trimmedContent = content.trim();
      onBlur && onBlur(index, trimmedContent === option.email ? option : { email: trimmedContent });
      event.stopPropagation();
      event.preventDefault();
    },
    [onBlur, index, option, content],
  );

  const onInputBlur = useCallback(
    (event) => {
      if (event.relatedTarget) {
        onTagBlur(event);
      }
    },
    [onTagBlur],
  );

  const onKeyDown = useCallback(
    (event) => {
      const trimmedContent = content.trim();

      switch (event.key) {
        case ENTER: {
          const newOption =
            selectedSuggestion && validateSuggestion(trimmedContent, selectedSuggestion)
              ? selectedSuggestion
              : trimmedContent === option.email
              ? option
              : { email: trimmedContent };

          onFinish && onFinish(index, newOption);
          event.preventDefault();
          break;
        }

        case COMMA:
        case SEMI:
          onFinish && onFinish(index, trimmedContent === option.email ? option : { email: trimmedContent });
          event.preventDefault();
          break;

        case ESCAPE:
          if (hasValidSuggestions) {
            setAutocompleteOpen(false);
            event.stopPropagation();
            break;
          }

          onFinish && onFinish(index, option);
          break;

        case BACKSPACE:
          if (!event.repeat && content === '' && onRemove) {
            onRemove(index);
          }
          break;

        case UP_ARROW:
          setSelectedSuggestion(moveSuggestion(validSuggestions, selectedSuggestion, -1));
          event.preventDefault();
          break;

        case DOWN_ARROW:
          setSelectedSuggestion(moveSuggestion(validSuggestions, selectedSuggestion, 1));
          event.preventDefault();
          break;

        default:
          if (!autocompleteOpen) {
            setAutocompleteOpen(true);
          }
      }
    },
    [
      onFinish,
      index,
      content,
      option,
      hasValidSuggestions,
      onRemove,
      validSuggestions,
      selectedSuggestion,
      autocompleteOpen,
    ],
  );

  const onSuggestionClick = useCallback(
    (suggestion, event) => {
      onFinish && onFinish(index, suggestion);
      event.stopPropagation();
      event.preventDefault();
    },
    [index, onFinish],
  );

  const onSuggestionHover = useCallback(
    (suggestion) => {
      setSelectedSuggestion(suggestion);
    },
    [setSelectedSuggestion],
  );

  if (editing) {
    return (
      <>
        {hasValidSuggestions && <Overlay active onOverlayClick={onTagBlur} backdrop="" />}
        <Box
          contentEditable
          suppressContentEditableWarning
          tabIndex={0}
          className={cx(theme['label--editing'], uiTypography['text'], uiTypography['text-body'])}
          ref={ref}
          onInput={onChange}
          onKeyDown={onKeyDown}
          onBlur={hasValidSuggestions ? onInputBlur : onTagBlur}
        >
          {option.email}
        </Box>
        {hasValidSuggestions && (
          <Autocomplete
            suggestions={validSuggestions}
            highlightedSuggestion={selectedSuggestion}
            onClick={onSuggestionClick}
            onHover={onSuggestionHover}
            renderSuggestion={renderSuggestion}
          />
        )}
      </>
    );
  }

  return (
    <Tag
      className={invalid ? theme['label--error'] : undefined}
      margin={1}
      marginRight={2}
      onFocus={onTagFocus}
      onRemoveClick={onRemove && onTagRemove}
    >
      <Link element="button" className={theme['label-text']} inherit={false} onClick={onTagClick}>
        {option.label || option.email}
      </Link>
    </Tag>
  );
};

const emailOption = PropTypes.shape({
  email: PropTypes.string.isRequired,
  label: PropTypes.string,
  id: PropTypes.string,
});

Label.propTypes = {
  option: emailOption.isRequired,
  index: PropTypes.number.isRequired,
  editing: PropTypes.bool,
  invalid: PropTypes.bool,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  onRemove: PropTypes.func,
  onFinish: PropTypes.func,
  suggestions: PropTypes.oneOfType([PropTypes.arrayOf(emailOption), PropTypes.object]),
  renderSuggestion: PropTypes.elementType,
};

export default Label;
