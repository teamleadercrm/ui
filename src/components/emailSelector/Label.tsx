import React, { useCallback, useState, useEffect, useRef, useMemo } from 'react';
import uiTypography from '@teamleader/ui-typography';
import cx from 'classnames';

import theme from './theme.css';
import { filterSuggestions, moveSuggestion, selectContentEditable, validateSuggestion } from './utils';
import Tag from '../tag';
import Link from '../link';
import Box from '../box';
import Autocomplete from './Autocomplete';
import Overlay from '../overlay';
import { Suggestion, Suggestions } from './types';
import EmailSuggestion from './EmailSuggestion';

const ENTER = 'Enter';
const ESCAPE = 'Escape';
const BACKSPACE = 'Backspace';
const UP_ARROW = 'ArrowUp';
const DOWN_ARROW = 'ArrowDown';
const SEMI = ';';
const COMMA = ',';

interface LabelProps {
  option: Suggestion;
  index: number;
  editing?: boolean;
  invalid?: boolean;
  onClick?: (index: number) => void;
  onFocus?: (index: number) => void;
  onBlur?: (index: number, newLabel: Suggestion | { email: string }) => void;
  onRemove?: (index: number) => void;
  onFinish?: (index: number, newLabel: Suggestion | { email: string }) => void;
  suggestions?: Suggestions;
  renderSuggestion?: React.ComponentType<React.ComponentPropsWithoutRef<typeof EmailSuggestion>>;
}

const Label = ({
  option,
  suggestions = [],
  editing,
  invalid,
  index,
  onClick,
  onFocus,
  onBlur,
  onRemove,
  onFinish,
  renderSuggestion,
}: LabelProps) => {
  const [content, setContent] = useState(option.email);
  const [autocompleteOpen, setAutocompleteOpen] = useState(true);
  const ref = useRef<HTMLDivElement>();

  const validSuggestions = useMemo(() => filterSuggestions(content, suggestions), [content, suggestions]);
  const [selectedSuggestion, setSelectedSuggestion] = useState<Suggestion | null>(null);

  const hasValidSuggestions =
    autocompleteOpen && (Array.isArray(validSuggestions) ? validSuggestions : Object.keys(validSuggestions)).length > 0;

  useEffect(() => {
    if (editing) {
      setContent(option.email);
      ref.current && selectContentEditable(ref.current);
    }
  }, [editing, option]);

  const onTagClick = useCallback(
    (event) => {
      onClick && onClick(index);
      event.stopPropagation();
      event.preventDefault();
    },
    [index, onClick],
  );

  const onTagFocus = useCallback(() => {
    onFocus && onFocus(index);
  }, [index, onFocus]);

  const onTagRemove = useCallback(
    (event) => {
      onRemove && onRemove(index);
      event.stopPropagation();
      event.preventDefault();
    },
    [index, onRemove],
  );

  const onChange = useCallback(
    (event) => {
      setContent(event.target.innerText);
    },
    [setContent],
  );

  const onTagBlur = useCallback(
    (event) => {
      const trimmedContent = content.trim();
      onBlur && onBlur(index, trimmedContent === option.email ? option : { email: trimmedContent });
      event.stopPropagation();
      event.preventDefault();
    },
    [onBlur, index, option, content],
  );

  const onInputBlur = useCallback(
    (event) => {
      if (event.relatedTarget) {
        onTagBlur(event);
      }
    },
    [onTagBlur],
  );

  const onKeyDown = useCallback(
    (event) => {
      const trimmedContent = content.trim();

      switch (event.key) {
        case ENTER: {
          const newOption =
            selectedSuggestion && validateSuggestion(trimmedContent, selectedSuggestion)
              ? selectedSuggestion
              : trimmedContent === option.email
              ? option
              : { email: trimmedContent };

          onFinish && onFinish(index, newOption);
          event.preventDefault();
          break;
        }

        case COMMA:
        case SEMI:
          onFinish && onFinish(index, trimmedContent === option.email ? option : { email: trimmedContent });
          event.preventDefault();
          break;

        case ESCAPE:
          if (hasValidSuggestions) {
            setAutocompleteOpen(false);
            event.stopPropagation();
            break;
          }

          onFinish && onFinish(index, option);
          break;

        case BACKSPACE:
          if (!event.repeat && content === '' && onRemove) {
            onRemove(index);
          }
          break;

        case UP_ARROW:
          setSelectedSuggestion(moveSuggestion(validSuggestions, selectedSuggestion, -1));
          event.preventDefault();
          break;

        case DOWN_ARROW:
          setSelectedSuggestion(moveSuggestion(validSuggestions, selectedSuggestion, 1));
          event.preventDefault();
          break;

        default:
          if (!autocompleteOpen) {
            setAutocompleteOpen(true);
          }
      }
    },
    [
      onFinish,
      index,
      content,
      option,
      hasValidSuggestions,
      onRemove,
      validSuggestions,
      selectedSuggestion,
      autocompleteOpen,
    ],
  );

  const onSuggestionClick = useCallback(
    (suggestion, event) => {
      onFinish && onFinish(index, suggestion);
      event.stopPropagation();
      event.preventDefault();
    },
    [index, onFinish],
  );

  const onSuggestionHover = useCallback(
    (suggestion) => {
      setSelectedSuggestion(suggestion);
    },
    [setSelectedSuggestion],
  );

  if (editing) {
    return (
      <>
        {hasValidSuggestions && (
          <Overlay active onOverlayClick={onTagBlur} backdrop="" className={theme['label-overlay']} />
        )}
        <Box
          contentEditable
          suppressContentEditableWarning
          tabIndex={0}
          className={cx(theme['label--editing'], uiTypography['text'], uiTypography['text-body'])}
          ref={ref}
          onInput={onChange}
          onKeyDown={onKeyDown}
          onBlur={hasValidSuggestions ? onInputBlur : onTagBlur}
        >
          {option.email}
        </Box>
        {hasValidSuggestions && (
          <Autocomplete
            suggestions={validSuggestions}
            highlightedSuggestion={selectedSuggestion}
            onClick={onSuggestionClick}
            onHover={onSuggestionHover}
            renderSuggestion={renderSuggestion}
          />
        )}
      </>
    );
  }

  return (
    <Tag
      className={invalid ? theme['label--error'] : undefined}
      margin={1}
      marginRight={2}
      onFocus={onTagFocus}
      onRemoveClick={onRemove && onTagRemove}
    >
      <Link element="button" className={theme['label-text']} inherit={false} onClick={onTagClick}>
        {option.label || option.email}
      </Link>
    </Tag>
  );
};

export default Label;
