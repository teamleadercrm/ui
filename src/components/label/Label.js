import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import Box from '../box';
import { TextBodyCompact, TextDisplay, TextSmall } from '../typography';

export default class Label extends PureComponent {
  render() {
    const { children, connectedLeft, connectedRight, inverse, helpText, required, size, ...others } = this.props;

    const childProps = {
      inverse,
      marginTop: 1,
      size,
    };

    const Element = size === 'large' ? TextDisplay : TextBodyCompact;

    return (
      <Box display="block" element="label" marginBottom={3} {...others}>
        {React.Children.map(children, (child) => {
          if (typeof child !== 'string') {
            return React.cloneElement(child, { ...childProps, ...child.props });
          }

          return (
            <Box display="flex" alignItems="center">
              {connectedLeft && (
                <Box element="span" marginRight={1}>
                  {connectedLeft}
                </Box>
              )}
              <Element color={inverse ? 'neutral' : 'teal'} tint={inverse ? 'lightest' : 'darkest'} element="span">
                {child}
              </Element>
              {!required && (
                <TextSmall
                  color={inverse ? 'teal' : 'neutral'}
                  element="span"
                  marginLeft={1}
                  tint={inverse ? 'light' : 'darkest'}
                >
                  {helpText}
                </TextSmall>
              )}
              {connectedRight && (
                <Box element="span" marginLeft={1}>
                  {connectedRight}
                </Box>
              )}
            </Box>
          );
        })}
      </Box>
    );
  }
}

Label.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.string, PropTypes.array]),
  connectedLeft: PropTypes.element,
  connectedRight: PropTypes.element,
  inverse: PropTypes.bool,
  helpText: PropTypes.string,
  required: PropTypes.bool,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
};

Label.defaultProps = {
  inverse: false,
  helpText: 'Optional',
  required: false,
  size: 'medium',
};
