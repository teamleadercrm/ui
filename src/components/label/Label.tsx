import { IconInfoBadgedSmallFilled } from '@teamleader/ui-icons';
import React, { ReactNode } from 'react';
import { GenericComponent } from '../../@types/types';
import { SIZES } from '../../constants';
import Box from '../box';
import { BoxProps } from '../box/Box';
import Icon from '../icon';
import Tooltip from '../tooltip';
import { TextBodyCompact, TextDisplay, TextSmall } from '../typography';

const TooltippedIcon = Tooltip(Icon);

export interface LabelProps extends Omit<BoxProps, 'children'> {
  children: ReactNode;
  inverse?: boolean;
  required?: boolean;
  size?: Exclude<typeof SIZES[number], 'fullscreen' | 'smallest' | 'hero'>;
  tooltip?: ReactNode;
  tooltipProps?: Record<string, any>;
}

const Label: GenericComponent<LabelProps> = ({
  children,
  inverse = false,
  required = false,
  size = 'medium',
  tooltip,
  tooltipProps,
  ...others
}) => {
  const childProps = {
    inverse,
    marginTop: 1,
    size,
  };

  const Element = {
    tiny: TextSmall,
    small: TextBodyCompact,
    medium: TextBodyCompact,
    large: TextDisplay,
  }[size];

  return (
    <Box data-teamleader-ui="label" display="block" element="label" marginBottom={3} {...others}>
      {React.Children.map(children, (child) =>
        typeof child !== 'string' && React.isValidElement(child) ? (
          React.cloneElement(child, { ...childProps, ...child.props })
        ) : (
          <Box display="flex" alignItems="center">
            <Element color={inverse ? 'neutral' : 'teal'} tint={inverse ? 'lightest' : 'darkest'} element="span">
              {child}
            </Element>
            {required && (
              <TextSmall color="ruby" element="span" marginLeft={1} tint="dark">
                *
              </TextSmall>
            )}
            {tooltip && (
              <TooltippedIcon
                tooltip={<TextSmall>{tooltip}</TextSmall>}
                tooltipSize="small"
                color={inverse ? 'neutral' : 'teal'}
                tint={inverse ? 'lightest' : 'darkest'}
                marginLeft={1}
                {...tooltipProps}
              >
                <IconInfoBadgedSmallFilled />
              </TooltippedIcon>
            )}
          </Box>
        ),
      )}
    </Box>
  );
};

export default Label;
