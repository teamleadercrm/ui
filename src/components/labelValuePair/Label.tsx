import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Heading5 } from '../typography';

class Label extends PureComponent {
  render() {
    const { children, inline, ...others } = this.props;

    return (
      <Heading5
        color="teal"
        flex={inline ? '0 0 40%' : 1}
        marginBottom={inline ? 0 : 1}
        marginRight={inline ? 2 : 0}
        maxLines={2}
        paddingVertical={inline ? 1 : 0}
        {...others}
      >
        {children}
      </Heading5>
    );
  }
}

Label.propTypes = {
  children: PropTypes.node,
  inline: PropTypes.bool,
};

export default Label;
