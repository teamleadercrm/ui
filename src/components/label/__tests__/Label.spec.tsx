import React from 'react';
import { render } from '@testing-library/react';
import Label from '../Label';

describe('Component - Label', () => {
  it('renders', async () => {
    const { asFragment } = render(<Label required>Foobar</Label>);
    expect(asFragment()).toMatchSnapshot();
  });
});
