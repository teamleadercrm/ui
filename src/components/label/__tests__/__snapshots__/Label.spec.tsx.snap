// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Label renders 1`] = `
<DocumentFragment>
  <label
    class="box display-block margin-bottom-3"
    data-teamleader-ui="label"
  >
    <div
      class="box align-items-center display-flex"
    >
      <span
        class="box text text-body-compact darkest teal"
        data-teamleader-ui="text"
      >
        Foobar
      </span>
      <span
        class="box margin-left-1 text text-small dark ruby"
        data-teamleader-ui="text"
      >
        *
      </span>
    </div>
  </label>
</DocumentFragment>
`;
